protobuf/decode.go

```diff
@@ -0,0 +1,282 @@
+			if err != nil {
+				return nil, fmt.Errorf("Error decoding account storage entry; %w", err)
+			}
+			world[address].Storage[key] = value
+		}
+	}
+
+	return &world, nil
+}
+
+func (entry *Substate_AllocEntry) decode() ([]byte, *Substate_Account, error) {
+	return entry.GetAddress(), entry.GetAccount(), nil
+}
+
+func (acct *Substate_Account) decode() (uint64, *big.Int, []byte, types.Hash, error) {
+	return acct.GetNonce(),
+		types.BytesToBigInt(acct.GetBalance()),
+		acct.GetCode(),
+		types.BytesToHash(acct.GetCodeHash()),
+		nil
+}
+
+func (entry *Substate_Account_StorageEntry) decode() (types.Hash, types.Hash, error) {
+	return types.BytesToHash(entry.GetKey()),
+		types.BytesToHash(entry.GetValue()),
+		nil
+}
+
+// decode converts protobuf-encoded Substate_BlockEnv into aida-comprehensible Env
+func (env *Substate_BlockEnv) decode() (*substate.Env, error) {
+	var blockHashes map[uint64]types.Hash = nil
+	if env.GetBlockHashes() != nil {
+		blockHashes = make(map[uint64]types.Hash, len(env.GetBlockHashes()))
+		for _, entry := range env.GetBlockHashes() {
+			key, value, err := entry.decode()
+			if err != nil {
+				return nil, err
+			}
+			blockHashes[key] = types.BytesToHash(value)
+		}
+	}
+
+	return &substate.Env{
+		Coinbase:    types.BytesToAddress(env.GetCoinbase()),
+		Difficulty:  types.BytesToBigInt(env.GetDifficulty()),
+		GasLimit:    env.GetGasLimit(),
+		Number:      env.GetNumber(),
+		Timestamp:   env.GetTimestamp(),
+		BlockHashes: blockHashes,
+		BaseFee:     BytesValueToBigInt(env.GetBaseFee()),
+		Random:      BytesValueToHash(env.GetRandom()),
+		BlobBaseFee: BytesValueToBigInt(env.GetBlobBaseFee()),
+	}, nil
+}
+
+func (entry *Substate_BlockEnv_BlockHashEntry) decode() (uint64, []byte, error) {
+	return entry.GetKey(), entry.GetValue(), nil
+}
+
+// decode converts protobuf-encoded Substate_TxMessage into aida-comprehensible Message
+func (msg *Substate_TxMessage) decode(lookup dbGetCode) (*substate.Message, error) {
+
+	// to=nil means contract creation
+	var pTo *types.Address = nil
+	to := msg.GetTo()
+	if to != nil {
+		address := types.BytesToAddress(to.GetValue())
+		pTo = &address
+	}
+
+	// if InitCodeHash exists:
+	// 1. code = lookup the code using InitCodeHash
+	// 2. set data -> code from (1)
+	// 3. clear InitCodeHash
+	var data []byte = msg.GetData()
+	if pTo == nil {
+		code, err := lookup(types.BytesToHash(msg.GetInitCodeHash()))
+		if err != nil && !errors.Is(err, leveldb.ErrNotFound) {
+			return nil, fmt.Errorf("failed to decode tx message; %w", err)
+		}
+		data = code
+	}
+
+	txType := msg.GetTxType()
+
+	// Berlin hard fork, EIP-2930: Optional access lists
+	var accessList types.AccessList = nil // nil if EIP-2930 is not activated
+	switch txType {
+	case Substate_TxMessage_TXTYPE_ACCESSLIST,
+		Substate_TxMessage_TXTYPE_DYNAMICFEE,
+		Substate_TxMessage_TXTYPE_BLOB:
+
+		accessList = make([]types.AccessTuple, len(msg.GetAccessList()))
+		for i, entry := range msg.GetAccessList() {
```

Comment:
If `accessList` can be optional, it can be `nil` hence this can panic. Or is it just empty?

Reply:
`[]types.AccessTuple{}` is acceptable when `nil` is expected.